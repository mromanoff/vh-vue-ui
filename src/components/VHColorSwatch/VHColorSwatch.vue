<script type="text/jsx">
export default {
  name: 'VhColorSwatch',
  functional: true,
  props: {
    /**
     * Color Swatch will have active state
     */
    active: {
      type: Boolean,
      default: false,
    },
    /**
     * Color Swatch will take any valid CSS color value and apply it to background-color
     */
    color: {
      type: String,
      default: '',
      required: true,
    },
    /**
     * Color Swatch will be disabled
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Swatch will become either "small" or "medium"
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => {
        return value.match(/(small|medium)/);
      },
    },
  },

  render(h, { props, listeners }) {
    const classes = {
      ColorSwatch: true,
      'ColorSwatch--active': props.active,
      'ColorSwatch--disabled': props.disabled,
      [`ColorSwatch--${props.size}`]: props.size,
    };

    const innerClasses = {
      'ColorSwatch-inner': true,
      [`ColorSwatch-inner--${props.size}`]: props.size,
    };

    const styles = `background-color: ${props.color}`;

    return (
      <button class={classes} type="button" {...{ on: listeners }}>
        <span class={innerClasses} style={styles} />
      </button>
    );
  },
};
</script>

<style scoped>
.ColorSwatch {
  --ColorSwatch-outline-color: var(--color--grey500);
  --ColorSwatch-active-outline-color: var(--color--grey300);
  --ColorSwatch-size--base: 48px;
  --ColorSwatch-size--inner: 44px;
  --ColorSwatch-size--small: 24px;
  --ColorSwatch-size-inner--small: 22px;
  flex: 0 0 var(--ColorSwatch-size--base);
  align-items: center;
  justify-content: center;
  height: var(--ColorSwatch-size--base);
  width: var(--ColorSwatch-size--base);
  display: inline-flex;
  border-radius: 50%;
  outline: none;
  background-color: var(--color--blue100);
  cursor: pointer;
}

.ColorSwatch-inner {
  width: var(--ColorSwatch-size--inner);
  height: var(--ColorSwatch-size--inner);
  border-radius: 50%;
  display: block;
  border: 3px solid var(--color--white);
  transition: all 200ms ease-in-out;
}

.ColorSwatch-inner:hover {
  border: 3px solid var(--color--blue100);
}

.ColorSwatch--small {
  height: var(--ColorSwatch-size--small);
  width: var(--ColorSwatch-size--small);
  flex: 0 0 var(--ColorSwatch-size--small);
}

.ColorSwatch-inner--small {
  width: var(--ColorSwatch-size-inner--small);
  height: var(--ColorSwatch-size-inner--small);
  border: 2px solid var(--color--white);
  transition: all 200ms ease-in-out;
}

.ColorSwatch-inner--small:hover {
  border: 2px solid var(--color--blue100);
}

.ColorSwatch--active {
  pointer-events: none;
  background-color: var(--color--blue700);
}

.ColorSwatch--disabled {
  pointer-events: none;
}
</style>
