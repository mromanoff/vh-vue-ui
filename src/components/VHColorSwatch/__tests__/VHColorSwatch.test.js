import { shallowMount } from '@vue/test-utils';
import VHColorSwatch from '../VHColorSwatch.vue';

describe('VHColorSwatch.vue', () => {
  let wrapper;

  // TEST functional component with passing props: https://vue-test-utils.vuejs.org/api/options.html#context

  beforeEach(() => {
    wrapper = shallowMount(VHColorSwatch, {
      context: {
        props: { size: 'medium', color: '#ffcb59' },
      },
    });
  });

  it('renders default VHColorSwatch', () => {
    expect(wrapper.isFunctionalComponent).toBe(true);
    expect(wrapper.is('button')).toBe(true);
    expect(wrapper.classes()).toStrictEqual(['ColorSwatch', 'ColorSwatch--medium']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Disabled VHColorSwatch', () => {
    wrapper = shallowMount(VHColorSwatch, {
      context: {
        props: { disabled: true, size: 'medium', color: '#ffcb59' },
      },
    });
    expect(wrapper.is('button')).toBe(true);
    expect(wrapper.classes()).toStrictEqual(['ColorSwatch', 'ColorSwatch--disabled', 'ColorSwatch--medium']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Active VHColorSwatch', () => {
    wrapper = shallowMount(VHColorSwatch, {
      context: {
        props: { active: true, size: 'medium', color: '#ffcb59' },
      },
    });
    expect(wrapper.is('button')).toBe(true);
    expect(wrapper.classes()).toStrictEqual(['ColorSwatch', 'ColorSwatch--active', 'ColorSwatch--medium']);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Small VHColorSwatch', () => {
    wrapper = shallowMount(VHColorSwatch, {
      context: {
        props: { size: 'small', color: '#ffcb59' },
      },
    });
    expect(wrapper.is('button')).toBe(true);
    expect(wrapper.classes()).toStrictEqual(['ColorSwatch', 'ColorSwatch--small']);
    expect(wrapper).toMatchSnapshot();
  });
});
