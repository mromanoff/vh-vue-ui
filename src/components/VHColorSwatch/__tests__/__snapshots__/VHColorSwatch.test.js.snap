// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VHColorSwatch.vue renders Active VHColorSwatch 1`] = `<button type="button" class="ColorSwatch ColorSwatch--active ColorSwatch--medium"><span class="ColorSwatch-inner ColorSwatch-inner--medium" style="background-color: rgb(255, 203, 89);"></span></button>`;

exports[`VHColorSwatch.vue renders Disabled VHColorSwatch 1`] = `<button type="button" class="ColorSwatch ColorSwatch--disabled ColorSwatch--medium"><span class="ColorSwatch-inner ColorSwatch-inner--medium" style="background-color: rgb(255, 203, 89);"></span></button>`;

exports[`VHColorSwatch.vue renders Small VHColorSwatch 1`] = `<button type="button" class="ColorSwatch ColorSwatch--small"><span class="ColorSwatch-inner ColorSwatch-inner--small" style="background-color: rgb(255, 203, 89);"></span></button>`;

exports[`VHColorSwatch.vue renders default VHColorSwatch 1`] = `<button type="button" class="ColorSwatch ColorSwatch--medium"><span class="ColorSwatch-inner ColorSwatch-inner--medium" style="background-color: rgb(255, 203, 89);"></span></button>`;
